reasoner/planner/RecursivePlanner.java

```diff
@@ -0,0 +1,382 @@
+
+        // TODO: Combine these when we fix ReasonerPlanner.dependencies
+        PartialOrderReductionSearch(Set<Resolvable<?>> resolvables, Set<Variable> mode) {
+            this(resolvables, mode, ReasonerPlanner.dependencies(resolvables), true);
+        }
+
+        PartialOrderReductionSearch(Set<Resolvable<?>> resolvables, Set<Variable> mode, Map<Resolvable<?>, Set<Variable>> dependencies, boolean enableConnectednessRestriction) {
+            this.resolvables = resolvables;
+            this.mode = mode;
+            this.dependencies = dependencies;
+            this.enableConnectednessRestriction = enableConnectednessRestriction;
+        }
+
+        List<List<Resolvable<?>>> allOrderings() {
+            Set<Resolvable<?>> remaining = new HashSet<>(resolvables);
+            Set<Variable> bounds = new HashSet<>(mode);
+            iterate(resolvables).filter(resolvable -> !resolvable.isNegated()).flatMap(r -> iterate(estimateableVariables(r)))
+                    .filter(v -> iterate(v.constraints()).anyMatch(constraint ->
+                            constraint.isThing() && constraint.asThing().isValue() && constraint.asThing().asValue().isValueIdentity()))
+                    .forEachRemaining(bounds::add);
+
+            List<List<Resolvable<?>>> orderings = new ArrayList<>();
+            porDFS(new Stack<>(), bounds, remaining, new HashSet<>(), orderings);
+            return orderings; // TODO: If this causes a memory blow-up, POR can "generate" one at a time.
+        }
+
+        private void porDFS(Stack<Resolvable<?>> currentPath, Set<Variable> currentBounds,
+                            Set<Resolvable<?>> remaining, Set<Resolvable<?>> sleepSet, List<List<Resolvable<?>>> orderings) {
+            if (remaining.isEmpty()) {
+                orderings.add(new ArrayList<>(currentPath));
+                return;
+            }
+
+            List<Resolvable<?>> enabled = iterate(remaining)
+                    .filter(r -> ReasonerPlanner.dependenciesSatisfied(r, currentBounds, dependencies) && !sleepSet.contains(r))
+                    .toList();
+
+            if (enabled.isEmpty()) {
+                enabled = iterate(remaining).filter(r -> !r.isNegated() && !sleepSet.contains(r)).toList();
+            }
+
+            if (enabled.isEmpty()) {
+                return; // All enabled are sleeping
+            }
+
+            if (enableConnectednessRestriction) {
+                List<Resolvable<?>> connectedEnabled = iterate(enabled)
+                        .filter(r -> iterate(estimateableVariables(r)).anyMatch(currentBounds::contains)).toList();
+                if (!connectedEnabled.isEmpty()) {
+                    enabled = connectedEnabled;
+                }
+            }
+
+            Set<Resolvable<?>> newSleepSet = new HashSet<>(sleepSet);
+            for (Resolvable<?> next : enabled) {
+                Set<Resolvable<?>> awaken = iterate(newSleepSet) //  Don't need to iterate(remaining), do I?
```

Comment:
Make sure you resolve and remove this kind of comment. We should never let this kind of comment make it into production code!